### ISSUE TITLE
multiple metrics violate datadog best practices and impact billing

### ISSUE THREAD

**Issue by eedwards-sk** [concourse/concourse#3196]
## What challenge are you facing?

Currently, metric tags emitted to datadog for the `http_response_time` metric include tags for the `path` which when correlating to build numbers or other dynamic paths, may become an ever-increasing list of tags, impacting billing and cardinality limits

e.g.:

`path:/api/v1/teams/main/pipelines/long/jobs/long-running-process/builds/29`

in addition the `build_finished` and `build_started` metrics include the `build_id`

because each tag on a metric becomes a new dimension in datadog, this means that:

- each `path` or `build_id`, which at the very least will be unique for each build, means that each build results in a new custom metric
- datadog billing charges based on custom metric use, meaning every single build can potentially increase billing cost
- because a build number is a counter, this will also start to push datadog cardinality limits as build counts increase, e.g. to 1000+
- any other highly dynamic paths would also push this limit

https://docs.datadoghq.com/tagging/#defining-tags
https://docs.datadoghq.com/developers/metrics/custom_metrics/#how-is-a-custom-metric-defined

## What would make this better?

- Stop tagging `http_response_time` metric with the `path`

- Stop tagging `build_started` and `build_finished` metrics with the `build_id`, and any other metrics adding tags with 'counters'  or dynamic / ever increasing values.

- Alternatively, identify the most dynamic sections of the metrics and consolidate them, e.g.:

`path:/api/v1/teams/main/pipelines/long/jobs/long-running-process/builds/29`

could be reduced to

`path:/api/v1/teams/main/pipelines/long/jobs/long-running-process/builds`

but if a user is generating a lot of dynamic pipelines, jobs, or tasks, it would still possibly be an excessive usage of custom metrics

- Alternatively, allow a metric blacklist in the configuration so I can stop reporting these metrics.

## Are you interested in implementing this yourself?

Theoretically Yes, I will help however I can, but I doubt I can do the implementation myself with my current knowledge of Go and the codebase.

**Comment by eedwards-sk** [comment]
Based on my datadog billing page for usage, the top three custom metrics (by tag usage) from concourse are:

```
build_finished
build_started
http_response_time
```

**Comment by eedwards-sk** [comment]
suggesting we remove or make configurable the submission of...

`path`:

https://github.com/concourse/concourse/blob/33d51223f4baf3f571d5e536b8992504fe9ee364/metric/metrics.go#L206-L241

`build_id`:

https://github.com/concourse/concourse/blob/33d51223f4baf3f571d5e536b8992504fe9ee364/metric/metrics.go#L140-L200

**Comment by vito** [comment]
Yeah, I think we'll need to do something about these high-cardinality metrics.

Related: #2896

**Comment by gordlea** [comment]
It would also be nice if build_started and build_finished were events instead of a gauge metric. It's hard to do anything with it as is, as it seems like the metric value is the build id, which doesn't really mean anything on a chart.

**Comment by eedwards-sk** [comment]
update... datadog now supports concourse as a first-party integration

this means if you correctly configure your metric prefix to be `concourse.ci`, datadog will not consider concourse metrics as "custom metrics", and they won't impact billing

huzzah!

this still doesn't mean we shouldn't address the cardinality, because it definitely could have negative impact on their infrastructure as per their best practices, but it does significantly reduce the possibility of a severe impact to billing

**Comment by vito** [comment]
@eedwards-sk Woah, that's a surprise. This, right?: https://docs.datadoghq.com/integrations/concourse_ci/

Neat!

**Comment by eedwards-sk** [comment]
@vito Yep! concourse getting some serious clout out on the streets, yo! :D

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
